torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -434,27 +432,48 @@ bool ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const {
   return true;
 }
 
-void ProcessGroupNCCL::WorkNCCL::checkAndThrowException() {
-  // Set the appropriate exception if found.
-  checkAndSetException();
+bool ProcessGroupNCCL::WorkNCCL::checkTimeout() {
+  auto currentTimepoint = std::chrono::steady_clock::now();
+  auto timeElapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+      currentTimepoint - workStartTime_);
 
-  // Throw an exception, only if we have a valid exception.
-  if (exception()) {
-    std::rethrow_exception(exception());
-  }
+  if (timeElapsed < opTimeout_)
+    return false;
+
+  // Timed out
+
+  // There is already an error, we don't override it
+  if (exception())
+    return true;
```

Comment:
nit: might be useful to log that we hit an existing exception and what it is, and also mention that there was also indeed a timeout?